classes/file_search.php

```diff
@@ -171,23 +174,50 @@ public static function files(files $record, string $output = '', int $limitfrom
             $output = $dir . '/' . $record->get_temp_filename();
         }
 
-        $stream = fopen($output, 'w');
-        $columnheaders = [
-            'fileid', 'courseid', 'shortname', 'contextid', 'component', 'filearea', 'itemid', 'filepath', 'filename',
-            'mimetype', 'strategy', 'internal', 'replace', 'offset', 'match',
-        ];
-        fputcsv($stream, $columnheaders);
-
         [$whereclause, $params] = self::make_where_clause($criteria);
+        // If we are running a shard, then restrict the range of id.
+        if ( ! empty($startid) || ! empty($endid)) {
+            if (empty($finalshard)) {
+                $logmessage .= " Shard from $startid to $endid.";
+                $whereclause .= ' AND f.id between :startid and :endid';
+                $params['startid'] = $startid;
+                $params['endid'] = $endid;
+            } else {
+                $logmessage .= " Final shard from $startid.";
+                $whereclause .= ' AND f.id >= :startid';
+                $params['startid'] = $startid;
+            }
+        }
+
+        // If the output file already exists, try to resume.
+        if (file_exists($output)) {
+            // This must be a resumed job. We need to append to previous output.
+            [$resumeid, $matchcount] = self::resume($output);
+        } else {
+            $resumeid = 0;
+            $matchcount = 0;
+        }
+        if ( ! empty($resumeid)) {
+            $logmessage .= " Resume from $resumeid.";
+            $stream = fopen($output, 'a');
+            $whereclause .= ' AND f.id >= :resumeid ';
+            $params['resumeid'] = $resumeid;
+        } else {
+            $stream = fopen($output, 'w');
+            $columnheaders = [
+                'fileid', 'courseid', 'shortname', 'contextid', 'component', 'filearea', 'itemid', 'filepath', 'filename',
+                'mimetype', 'strategy', 'internal', 'replace', 'offset', 'match',
+            ];
+            fputcsv($stream, $columnheaders);
+        }
+
+        mtrace($logmessage);
         $record->set('timestart', time());
         $updatetime = time();
         $updatepercent = 0;
-        $matchcount = 0;
         $filecount = 0;
```

Comment:
We're resetting the startid, updatepercent and filecount for the resume. 

This will all work, it just resets progress back to 0. I think it's fine to leave this as is for now, but ideally we would calculate the filecount using the original percent, and also add this to totalfiles (this would also impact the ETA - which is why I'm leaning towards leaving it for later)

Reply:
I agree about leaving it for later.  Then, how about putting the totalfiles and filecount in the record?  Again, that is explicit coding.

Reply:
It's an option, but as this is just for progress and estimates are fine I don't think we actually need to be explicit here. timeresumed may be needed for the ETA though. But lets leave this for later.